p2panda-net/src/sync/manager.rs

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+
+use anyhow::Result;
+use iroh_gossip::proto::TopicId;
+use iroh_net::{Endpoint, NodeId};
+use p2panda_sync::traits::SyncProtocol;
+use p2panda_sync::SyncError;
+use tokio::sync::mpsc::Sender;
+use tracing::{debug, warn};
+
+use crate::engine::ToEngineActor;
+use crate::sync::{self, SYNC_CONNECTION_ALPN};
+
+#[derive(Debug)]
+pub struct SyncManager {
+    known_peer_topics: HashMap<NodeId, HashSet<TopicId>>,
+    active_sync_sessions: HashMap<TopicId, HashSet<NodeId>>,
+    completed_sync_sessions: HashMap<TopicId, HashSet<NodeId>>,
+    engine_actor_tx: Sender<ToEngineActor>,
+    endpoint: Endpoint,
+    sync_protocol: Arc<dyn for<'a> SyncProtocol<'a> + 'static>,
+}
+
+impl SyncManager {
+    pub fn new(
+        endpoint: Endpoint,
+        engine_actor_tx: Sender<ToEngineActor>,
+        sync_protocol: Arc<dyn for<'a> SyncProtocol<'a> + 'static>,
+    ) -> Self {
+        Self {
+            known_peer_topics: HashMap::new(),
+            active_sync_sessions: HashMap::new(),
+            completed_sync_sessions: HashMap::new(),
+            engine_actor_tx,
+            endpoint,
+            sync_protocol,
+        }
+    }
+
+    /// Respond to newly discovered peer topics by initiating a new connection if one is not
+    /// currently underway and a successful sync session has not already been completed.
+    pub async fn update_peer_topics(&mut self, peer: NodeId, topics: Vec<TopicId>) -> Result<()> {
+        debug!("updating peer topics in connection manager");
+
+        // Create a list of (previously) unknown topics that we might like to sync over.
+        let mut new_topics = Vec::new();
+        let known_topics = self.known_peer_topics.get(&peer);
+        if let Some(known_topics) = known_topics {
+            for topic in topics {
+                if !known_topics.contains(&topic) {
+                    new_topics.push(topic)
+                }
+            }
+        } else {
+            new_topics = topics
+        }
+
+        for topic in new_topics {
+            // Peers with whom we have active outbound sync sessions for this topic.
+            let active_peers = self.active_sync_sessions.entry(topic).or_default();
+
+            // Have we already completed a successful sync session with this peer?
+            let sync_complete = self
+                .completed_sync_sessions
+                .entry(topic)
+                .or_default()
+                .contains(&peer);
+
+            // Attempt connection in order to initiate a sync session.
+            if !active_peers.contains(&peer) && !sync_complete {
+                active_peers.insert(peer);
+                // @TODO: We might want to disentangle this call from the `update_peer_topics`
```

Comment:
One way to do this would be to make `connect()` a standalone function (not a method) and spawn it in a task. I'm not overly concerned about it for now but that could be a target for cleanup phase. I'd probably also want to add a sync connection duration timer so we can get a feeling for how long sync takes in various scenarios.